### ISSUE TITLE
Different `--value=end` for `-p 2020` vs `-p 2020 -Y` (in 1.20)

### ISSUE THREAD

**Issue by julian-klode** [simonmichael/hledger#1424]
The file below produces inconsistent `--value=end` results for `-p 2020` and `-p 2020 -Y`; I have not calculated which one is correct. It's a bit hard to say - given that the last price is before the January purchase, I'd probably say there were no gains on that purchase, only on the one in December:

```
D 1,000.00€
commodity 1,000.00€

2019-12-23 buy 200 IE00BF4RFH31
    assets:investments    200 "IE00BF4RFH31" @ 4.9785€
    assets:investments                        -995.70€

2020-01-22 buy 103 IE00BF4RFH31
    assets:investments    103 "IE00BF4RFH31" @ 5.033€
    assets:investments                       -518.40€

P 2019/12/30 "IE00BF4RFH31" 4.957€
P 2020/01/02 "IE00BF4RFH31" 4.938€

```

The outcomes are:

```
; $ hledger -f f bal -p 2020   --value=end  assets -Y
; Period-end value changes in 2020:
; 
;                    ||    2020 
; ====================++=========
; assets:investments || -13.59€ 
; --------------------++---------
;                    || -13.59€ 
; $ hledger -f f bal -p 2020   --value=end  assets 
;              -9.79€  assets:investments
; --------------------
;              -9.79€
```

**Comment by julian-klode** [comment]
So it seems that `-p 2020` only takes account the investment made in 2020, whereas `-p 2020 -Y` also takes into account the investment made in 2019 or something like it.

**Comment by simonmichael** [comment]
Thanks for the report. Here's a simpler reproducer:
```
2019-12-15
    (a)    1 A

2020-01-15
    (a)    1 A

P 2019-12-30 A 10 B
P 2020-01-02 A 20 B
```

hledger 1.19 picks valuation dates 2020/1/15 and then 2020/12/31, per docs, giving the same results:

```
$ hledger-1.19 bal -p 2020  --value=end --debug
valuation date: 2020-01-15
default valuation commodity for A: Just "B"
market price for A -> B: 20.00000000
                20 B  a
--------------------
                20 B
```
```
$ hledger-1.19 bal -p 2020 -Y --value=end --debug
Balance changes in 2020, valued at period ends:

valuation date: 2020-12-31
default valuation commodity for A: Just "B"
market price for A -> B: 20.00000000
   || 2020 
===++======
 a || 20 B 
---++------
   || 20 B 
```

hledger-1.20 picks 2020-01-15 and then 2020/12/31 but also (?) 2019/12/31, causing different results:

```
$ hledger-1.20 bal -p 2020  --value=end --debug
valuation date: 2020-01-15
default valuation commodity for A: Just "B"
shortest price path for A -> B:
 Just
   [ MarketPrice
     { mpdate = 2020 - 01 - 02, mpfrom = "A", mpto = "B", mprate = 20 }
   ]
                20 B  a
--------------------
                20 B
```
```
$ hledger-1.20 bal -p 2020 -Y --value=end --debug
Period-end value changes in 2020:

valuation date: 2020-12-31
default valuation commodity for A: Just "B"
shortest price path for A -> B:
 Just
   [ MarketPrice
     { mpdate = 2020 - 01 - 02, mpfrom = "A", mpto = "B", mprate = 20 }
   ]
valuation date: 2019-12-31
default valuation commodity for A: Just "B"
shortest price path for A -> B:
 Just
   [ MarketPrice
     { mpdate = 2019 - 12 - 30, mpfrom = "A", mpto = "B", mprate = 10 }
   ]
   || 2020 
===++======
 a || 30 B 
---++------
   || 30 B 
```

hledger master picks 2020-12-31 for the first (per docs) and does the same as 1.20 for the second, giving results like 1.20's:

```
$ hledger bal -p 2020  --value=end --debug
valuation date: 2020-12-31
default valuation commodity for A: Just "B"
shortest price path for A -> B:
 Just
   [ MarketPrice
     { mpdate = 2020 - 01 - 02, mpfrom = "A", mpto = "B", mprate = 20 }
   ]
                20 B  a
--------------------
                20 B
```
```
$ hledger bal -p 2020 -Y --value=end --debug
Period-end value changes in 2020:

valuation date: 2020-12-31
default valuation commodity for A: Just "B"
shortest price path for A -> B:
 Just
   [ MarketPrice
     { mpdate = 2020 - 01 - 02, mpfrom = "A", mpto = "B", mprate = 20 }
   ]
valuation date: 2019-12-31
default valuation commodity for A: Just "B"
shortest price path for A -> B:
 Just
   [ MarketPrice
     { mpdate = 2019 - 12 - 30, mpfrom = "A", mpto = "B", mprate = 10 }
   ]
   || 2020 
===++======
 a || 30 B 
---++------
   || 30 B 
```

**Comment by simonmichael** [comment]
PS a workaround for now @julian-klode: force the valuation date with `--value=2020-12-31`.

**Comment by Xitian9** [comment]
Just to clarify: is the behaviour in current master correct, or still unexpected?

**Comment by Xitian9** [comment]
Aha, I've figured this out. The correct answer per the docs is 30B, for the following reason:

The starting balance is 1A, which is valued at the beginning of 2020 as 10B (=10B per 1A on 2019-12-31).
At the end of 2020, the balance is 2A, which is valued at the end of 2020 as 40B (=20B per 1A on 2020-12-31).
The difference is 30B, which is what is reported in the period change report, starting with #1380, which was included in hledger-1.20.

However, there was a code branch which does the simpler ̈in 2020 there was a gain of 1A, which is valued at 20B on 2020-12-31’ calculation in the case of single-period reports. I think this issue indicates that this behaviour is a bug, and the more complicated calculation should be done even for single-period reports. I've added a fix in #1440.

**Comment by simonmichael** [comment]
@Xitian9 I haven't fully understood your comment above, but isn't the 30 result (still visible in master) wrong, because this is a balance change report constrained to 2020, which should only be considering 2020 postings (with no starting balance) ?

**Comment by simonmichael** [comment]
Ie bal should match reg here:
```
~/src/hledger$ hledger -f a.j reg -w80 -p 2020 --value=end
2020-01-15                      (a)                           20 B          20 B
~/src/hledger$ hledger -f a.j bal -N -p 2020 --value=end
                30 B  a
```

**Comment by Xitian9** [comment]
The behaviour is as I intended it, whether that's correct/best behaviour is still up for debate. In this context, the period change report gives you the differences between the periods of the corresponding historical report.

The guiding principle I had here was that, for all valuation types, you should be able to reconstruct any of the Period Change, Cumulative, or Historical reports from any other, so long as you know the historical start balance.

**Comment by simonmichael** [comment]
@Xitian9: Sorry for not understanding and giving earlier feedback, and I expect your reasoning is good at least internally. But I can't understand this external behaviour and I think most users won't either. 

One important invariant we've always had, and should preserve: you can see the postings that make up a balance in a balance report, by running register with similar arguments.

**Comment by Xitian9** [comment]
Okay, then we have a hairier issue than assumed. So we should be tracking changes in value not of the period-end balance, but of the sum of postings that we have seen added newly to this report. This would then measure the fluctuation in the total value of the account over and above any fluctuation due to historical postings. Am I correct?

Just to put it formally, the invariant I was using to determine the balance behaviour is as follows:
If x_0 is the historical balance, and x_i are totals of the the new postings in each period, c_i are the cumulative totals in each period, and h_i are the historical totals in each period.
```
c_i == scanl (+) 0 [x_1,x_2,…,x_i]
h_i == scanl (+) x_0 [x_1,x_2,…,x_i]
```
The invariant was that these equations should hold regardless of the valuation strategy used (including no valuation).

**Comment by simonmichael** [comment]
The current status does not fit in my head / available bandwidth, and I'm too confused to say a lot about it. 1.20 and 1.21 are different worlds at this point and switching between them while we discuss N issues is killing my brain. Perhaps we could remove some complexity and make progress by focussing on 1.20 first.

Currently on 1.20-branch I'm trying to roll back #1353/#1380, by merging your #1447. I'm having trouble merging or cherry-picking, could you help me out by rebasing it against latest 1.20-branch ?

Once that's done, I'll be able to see more clearly the state of 1.20 and we can discuss any remaining actions needed to get 1.20.3 out.

**Comment by Xitian9** [comment]
Once we're done sorting out 1.20.3, I feel like someone (maybe me) should draft a new section in the manual explaining in detail the proposed new behaviour of `--value=end`. Once we agree on what the new behaviour should be, we can work on implementing it and a sufficiently detail suite of tests.

**Comment by simonmichael** [comment]
@julian-klode's original issue is fixed in master and in 1.20-branch (consistent results with or without -Y), and my "30 B" issue above is fixed in 1.20-branch. We'll continue discussion of the latter in the context of master on #1353 probably. Closing this one.
